File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
def solution(fees, records):
    parking = []
    tlog = sorted(list(set([a for a in zip(*[list(rc.split()) for rc in records])][1])))
    tlog = [[a, 0] for a in tlog]
    for rc in records:
        time, car, io = map(str, rc.split())
        h, m = map(int, time.split(':'))
        if io == "IN":
            parking.append([car, h, m])
        else:
            for p in parking:
                if p[0] == car:
                    t = (h - p[1]) * 60 + m - p[2]
                    parking.remove(p)
                    break
            for tl in tlog:
                if tl[0] == car:
                    tl[1] += t
                    break
    while parking:
        car, h, m = parking.pop()
        for tl in tlog:
            if tl[0] == car:
                tl[1] += (23 - h) * 60 + 59 - m
    answer = []
    for tl in tlog:
        if tl[1] > fees[0]:
            tmp = tl[1] - fees[0]
            if tmp % fees[2] != 0:
                tmp += fees[2] - tmp % fees[2]
            answer.append(fees[1] + tmp / fees[2] * fees[3])
        else:
            answer.append(fees[1])
    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;

class Solution {
    
    private int getTime(String time) {
        return Integer.parseInt(time.substring(0, 2)) * 60
            + Integer.parseInt(time.substring(3, 5));
    }
    
    private int getFee(int[] fees, int time) {
        if (time <= fees[0]) return fees[1];
        return fees[1] + 
            (int)Math.ceil(((double)time - fees[0]) / fees[2]) * fees[3];
    }
    
    public int[] solution(int[] fees, String[] records) {
        Map<String, List<Integer>> map = new TreeMap<>();
        for (String record : records) {
            String[] tmp = record.split(" ");
            int time = getTime(tmp[0]);
            map.computeIfAbsent(tmp[1], k -> new ArrayList<>())
                .add(time);
        }
        List<Integer> res = new ArrayList<>();
        for (List<Integer> list : map.values()) {
            if (list.size() % 2 != 0)
                list.add(1439);
            int totalTime = 0;
            for (int i = 0; i < list.size(); i += 2)
                totalTime += list.get(i+1) - list.get(i);
            res.add(getFee(fees, totalTime));
        }
        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}